### PR TITLE
Rework slot validation to fix crash and provide better errors

### PR DESCRIPTION
**Pull request by marctem**
### Purpose of this PR
Fixes HUP bug https://fogbugz.unity3d.com/f/cases/1143546/

Adds consistent slot conversion validation to every node by checking input slots to ensure that all edges provide convertible values (in AMN.CalculateNodeHasError()). This means if incoming value types change (i.e. hooking a Matrix input to a multiply node, thereby changing its output type), the node with the invalid _input_ will err. This removes the custom logic from MultiplyNode that tested its _outputs_, which was backwards.

Removes AMN.ImplicitConversionExists(), which was overly restrictive and now redundant.

Also moves all the slot type conversion functions and validation to SlotValue.cs so it's more obvious what they're doing.

---
### Release Notes
* Resolve graph hang when connecting invalid slots, provide better error messaging

---
### Testing status
**Katana Tests**: In progress: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=sg%2Fmatrix-cast-crash&unity_branch=trunk&automation-tools_branch=master&sort=1-asc

**Manual Tests**: What did you do?
- [X] Opened test project + Run graphic tests locally
- Other: 
Connected a lot of nodes to check edge validation.

**Automated Tests**: 
Editor test graph to validate all edge connection combinations work/fail as expected.
image

---
### Overall Product Risks
**Technical Risk**: Low - fixes validation but it's possible something depended on the weird corner cases of the old method.

**Halo Effect**: Medium - touches AMN, the base class for every node

### CODE REVIEW

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 1]
```diff
@@ -733,11 +731,8 @@ public void ValidateGraph()
                     inputSlot = inputNode.FindInputSlot<MaterialSlot>(edge.inputSlot.slotId);
                 }
 
-                if (outputNode == null
-                    || inputNode == null
-                    || outputSlot == null
-                    || inputSlot == null
-                    || !outputSlot.IsCompatibleWith(inputSlot))
```
Comment (pbbastian) [inline com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs]: Was this removal intended?

File: com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs [hunk 2]
```diff
@@ -112,7 +112,9 @@ public override void AddDefaultProperty(PropertyCollector properties, Generation
                     property = new Vector1ShaderProperty();
                     break;
                 default:
-                    throw new ArgumentOutOfRangeException();
+                    // This shouldn't happen due to edge validation. But if it does,
+                    // just ignore it. Asserting here doesn't help anybody.
+                    return;
```
[the remark this answers is not in the document]
Comment (marctem) [reply com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs]: This is a tough call of "crash fast" vs "don't break the user's graph". Once an unhandled exception is thrown, the graph is unsaveable and the user might lose work. I'll put a Debug.LogError() back in, but I don't think an exception is the right answer here.

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 1 above)
Comment (marctem) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs]: Yes. The first 4 are still there, but the last check was incorrect (a frame behind) and so would remove perfectly valid connections! With the better error reporting of incompatible connections, we don't need to automatically remove broken ones.

**Comment by marctem** [comment]
> Be careful with making formatting changes in your PRs. It makes your diff bigger, so the PR becomes harder to review and backport. IMO formatting is better done in a separate PR, and then run on all the Shader Graph code instead.

Yeah, my bad.

File: TestProjects/ShaderGraph/Packages/manifest.json [hunk 3]
```diff
@@ -1,5 +1,8 @@
 {
   "dependencies": {
+    "com.unity.ide.rider": "1.0.4",
```
[the remark this answers is not in the document]
Comment (marctem) [reply TestProjects/ShaderGraph/Packages/manifest.json]: Whoops, will remove.

File: com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pbbastian) [reply com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs]: Ah yeah, good point. LogError sounds like a good middle ground :)

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 1 above)
Comment (pbbastian) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs]: Perfect 👍

**Comment by zadi0648** [comment]
Still on the fence if this PR is big enough to warrant a test doc 🤔 , anyhow...

**Comment by zadi0648** [comment]
Same not-so-helpful error for the second case mentioned in https://fogbugz.unity3d.com/f/cases/1143546/ (but no freeze)
NewFreezeError

**Comment by zadi0648** [comment]
It should also be noted that interestingly two of the three bugs no longer reproduce in the same way in master anymore.

https://fogbugz.unity3d.com/f/cases/1143546/ no longer freezes graph, instead gets warning about Normal not accepting matrix.

2nd case in https://fogbugz.unity3d.com/f/cases/1143546/ no longer freezes, gets `Shader error in 'Shader Graphs/SavingFreeze': 'SG_FreezerSubGInOut_ab9ca51241ba62b4684d70d039e4c3cb': cannot implicitly convert from 'const float2x2' to 'float4x4' at line 223 (on d3d11)` in console
